extensions-core/kafka-extraction-namespace/src/main/java/io/druid/query/lookup/KafkaLookupExtractorFactory.java

```diff
@@ -98,220 +97,207 @@ public String fromBytes(byte[] bytes)
-                }
-              }
+  private Runnable makeRunnable(final Properties kafkaProperties) {
+    return new Runnable()
+    {
+      @Override
+      public void run()
+      {
+        while (!shouldStop()) {
+          // TODO: why is a new connector created and destroyed on every iteration?
+          final ConsumerConnector consumerConnector = consumerConnectorFactory.buildConnector(kafkaProperties);
+          try {
+            final List<KafkaStream<String, String>> streams = consumerConnector.createMessageStreamsByFilter(
+                new Whitelist(Pattern.quote(kafkaTopic)), 1, DEFAULT_STRING_DECODER, DEFAULT_STRING_DECODER
+            );
+
+            if (streams == null || streams.isEmpty()) {
+              throw new IAE("Topic [%s] had no streams", kafkaTopic);
             }
-          }
-      );
-      Futures.addCallback(
-          future, new FutureCallback<Object>()
-          {
-            @Override
-            public void onSuccess(Object result)
-            {
-              LOG.debug("Success listening to [%s]", topic);
+            if (streams.size() > 1) {
+              throw new ISE("Topic [%s] has %d streams! expected 1", kafkaTopic, streams.size());
             }
+            final KafkaStream<String, String> kafkaStream = streams.get(0);
 
-            @Override
-            public void onFailure(Throwable t)
-            {
-              if (t instanceof CancellationException) {
-                LOG.debug("Topic [%s] cancelled", topic);
-              } else {
-                LOG.error(t, "Error in listening to [%s]", topic);
+            // FIXME: Why is this done on every iteration?
+            startingReads.countDown();
+
+            for (final MessageAndMetadata<String, String> messageAndMetadata : kafkaStream) {
+              final String key = messageAndMetadata.key();
+              final String message = messageAndMetadata.message();
+              if (key == null || message == null) {
+                LOG.error("Bad key/message from topic [%s]: [%s]", kafkaTopic, messageAndMetadata);
+                continue;
               }
+              onNewEntry(key, message);
             }
-          },
-          MoreExecutors.sameThreadExecutor()
-      );
-      this.future = future;
-      final Stopwatch stopwatch = Stopwatch.createStarted();
-      try {
-        while (!startingReads.await(100, TimeUnit.MILLISECONDS) && connectTimeout > 0L) {
-          // Don't return until we have actually connected
-          if (future.isDone()) {
-            future.get();
-          } else {
-            if (stopwatch.elapsed(TimeUnit.MILLISECONDS) > connectTimeout) {
-              throw new TimeoutException("Failed to connect to kafka in sufficient time");
-            }
+          }
+          catch (Exception e) {
+            LOG.error(e, "Error reading stream for topic [%s]", kafkaTopic);
+          } finally {
+            Thread.interrupted();
```

Comment:
If we cancelled the future, the thread will be interrupted. However, the `consumerConnector.shutdown()` below will fail if the interrupted flag is up. I would remove the future cancellation altogether, but there is once usage of it that I do not understand yet, see the relevant comment